### PR TITLE
feat(currency_conversion): add currency conversion feature

### PR DESCRIPTION
**Pull request by prajjwalkumar17**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

A whole new crate for Currency Conversion which takes foreign exchange rates and converts a from_currency to to_currency relatively i.e via a base_currency.

Forex rates can be fetched using the route /forexrates which for now uses [OpenExchange](https://openexchangerates.org/) Api. Moreover we have a fallback api as well for cases in which our primary api fails.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?

Written Test cases in the conversion.rs for all the scenarios.

`curl --location 'http://127.0.0.1:8080/forex/rates' \
--header 'api-key: dev_YW404vk3R5qi4FwvkY0pQDYwwgv4LdDBPMlQvei44yZn4Zu2eIa6kHD9zwO5stSj' \
--data ''`

This will allow the route `forex/rates` to work .
Screenshot 2023-08-31 at 12 14 25

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/currency_conversion/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "currency_conversion"
+description = "Currency conversion for cost based routing"
+version = "0.1.0"
+edition.workspace = true
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[dependencies]
+# First party crates
+api_models = { version = "0.1.0", path = "../api_models", package = "api_models" }
```
Comment (vspecky) [inline crates/currency_conversion/Cargo.toml:10]: I think `common_enums` should suffice for this crate. Use `common_enums` instead of `api_models`

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

File: config/development.toml
Comment (SanchithHegde) [inline config/development.toml:1]: Please make this change in the remaining config files: Docker Compose config file, example config file with comments, and loadtest config file.

File: crates/router/src/utils/currency.rs [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+                return Ok(rates.clone());
+            }
+            Ok(None) => continue,
+            Err(e) => {
+                logger::error!(?e);
+                continue;
+            }
+        }
+    }
+    //acquire lock one last time and try to fetch and update local&redis
```
Comment (SanchithHegde) [inline crates/router/src/utils/currency.rs:146]:
```suggestion
    //acquire lock one last time and try to fetch and update local & redis
```

File: crates/router/src/utils/currency.rs [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+        None => {
+            // redis expired
+            successive_fetch_and_save_forex(
+                state,
+                Some(redis_data),
+                #[cfg(feature = "kms")]
+                kms_config,
+            )
+            .await
+            // fallback_forex_fetch_expired_redis(state, redis_data).await
```
Comment (SanchithHegde) [inline crates/router/src/utils/currency.rs:346]: Remove commented code

File: crates/router/src/utils/currency.rs [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+            }
+            None => {
+                logger::error!("Rates for {} not received from API", &enum_curr);
+            }
+        };
+    }
+
+    let rates =
+        FxExchangeRatesCacheEntry::new(ExchangeRates::new(enums::Currency::USD, conversions));
+    // save_forex_to_local(rates);
```
Comment (SanchithHegde) [inline crates/router/src/utils/currency.rs:526]: Remove commented code.

**Review by SanchithHegde — DISMISSED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by vspecky — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Pull request merged**